include/swift/AST/SubstitutionMap.h

```diff
@@ -55,48 +44,150 @@ enum class CombineSubstitutionMaps {
+  /// replacement types and conformances.
+  class Storage final
+    : llvm::TrailingObjects<Storage, Type, ProtocolConformanceRef>
+  {
+    friend TrailingObjects;
+
+    /// The generic signature for which we are performing substitutions.
+    GenericSignature * const genericSig;
+
+    /// The number of conformance requirements, cached to avoid constantly
+    /// recomputing it on conformance-buffer access.
+    const unsigned numConformanceRequirements;
+
+    Storage() = delete;
+
+    Storage(GenericSignature *genericSig,
+            ArrayRef<Type> replacementTypes,
+            ArrayRef<ProtocolConformanceRef> conformances);
+
+  private:
+    unsigned getNumReplacementTypes() const {
+      return genericSig->getGenericParams().size();
+    }
+
+    size_t numTrailingObjects(OverloadToken<Type>) const {
+      return getNumReplacementTypes();
+    }
+
+    size_t numTrailingObjects(OverloadToken<ProtocolConformanceRef>) const {
+      return numConformanceRequirements;
+    }
+
+  public:
+    /// Form storage for the given generic signature and its replacement
+    /// types and conformances.
+    static Storage *get(GenericSignature *genericSig,
+                        ArrayRef<Type> replacementTypes,
+                        ArrayRef<ProtocolConformanceRef> conformances);
+
+    /// Retrieve the generic signature that describes the shape of this
+    /// storage.
+    GenericSignature *getGenericSignature() const { return genericSig; }
+
+    /// Retrieve the array of replacement types, which line up with the
+    /// generic parameters.
+    ///
+    /// Note that the types may be null, for cases where the generic parameter
+    /// is concrete but hasn't been queried yet.
+    ArrayRef<Type> getReplacementTypes() const {
+      return llvm::makeArrayRef(getTrailingObjects<Type>(),
+                                getNumReplacementTypes());
+    }
+
+    MutableArrayRef<Type> getReplacementTypes() {
+      return MutableArrayRef<Type>(getTrailingObjects<Type>(),
+                                   getNumReplacementTypes());
+    }
+
+    /// Retrieve the array of protocol conformances, which line up with the
+    /// requirements of the generic signature.
+    ArrayRef<ProtocolConformanceRef> getConformances() const {
+      return llvm::makeArrayRef(getTrailingObjects<ProtocolConformanceRef>(),
+                                numConformanceRequirements);
+    }
+    MutableArrayRef<ProtocolConformanceRef> getConformances() {
+      return MutableArrayRef<ProtocolConformanceRef>(
+                                getTrailingObjects<ProtocolConformanceRef>(),
+                                numConformanceRequirements);
+    }
+  };
+
+  /// The storage needed to describe the set of substitutions.
+  ///
+  /// When null, this substitution map is empty, having neither a generic
+  /// signature nor any replacement types/conformances.
+  std::shared_ptr<Storage> storage;
```

Comment:
Is this a `shared_ptr` because it's nice to be able to freely copy and `SubstitutionMap` are often transient so allocating in the global arena is going to create a lot of junk?

Reply:
Oh, wait, you preempted this question in your commit message. Cool.